src/phoenix/metrics/retrieval_metrics.py

```diff
@@ -0,0 +1,103 @@
+from dataclasses import dataclass, field
+from typing import Iterable, Optional, cast
+
+import numpy as np
+import pandas as pd
+from sklearn.metrics import ndcg_score
+
+
+@dataclass(frozen=True)
+class RetrievalMetrics:
+    """
+    Ranking metrics computed on a list of evaluation scores sorted from high to
+    low by their ranking scores (prior to evaluation). For example, if the items
+    are search results and the evaluation scores are their relevance scores (e.g.
+    1 if relevant and 0 if not relevant), then the evaluation scores should be
+    sorted by the original order of the displayed results, i.e. the first search
+    result should go first. For more info on these metrics,
+    see https://cran.r-project.org/web/packages/recometrics/vignettes/Evaluating_recommender_systems.html
+    """  # noqa: E501
+
+    eval_scores: "pd.Series[float]"
+    length: int = field(init=False)
+    has_nan: bool = field(init=False)
+
+    def __init__(self, eval_scores: Iterable[float]) -> None:
+        _eval_scores = np.fromiter(eval_scores, dtype=float)
+        object.__setattr__(self, "length", len(_eval_scores))
+        object.__setattr__(self, "has_nan", not np.all(np.isfinite(_eval_scores)))
+        if self.length < 2:
+            # len < 2 won't work for sklearn.metrics.ndcg_score, so we pad it
+            # with zeros (but still keep track of the original length)
+            _scores = _eval_scores
+            _eval_scores = np.zeros(2)
+            _eval_scores[: len(_scores)] = _scores
+        # For ranking metrics, the actual scores used for ranking are only
+        # needed for sorting the items. Since we assume the items are already
+        # sorted from high to low by their ranking scores, we can assign ranking
+        # scores to be the reverse of the indices of eval_scores, just so that
+        # it goes from high to low.
+        ranking_scores = reversed(range(len(_eval_scores)))
+        object.__setattr__(
+            self,
+            "eval_scores",
+            pd.Series(_eval_scores, dtype=float, index=ranking_scores),  # type: ignore
+        )
+
+    def ndcg(self, k: Optional[int] = None) -> float:
```

Comment:
I think in this case we should side with mathematical purity and keep the return value as a `float`, because on one hand what we want to do next is aggregation and not having a uniform `float` output would complicate things, and on the other hand, leaving `k` as an unspecified value at call time is a way for callers to communicate that they do not care about `k`, so returning `k` is inconsistent with that intention (especially since that for these specific metrics, `k` only matters up to a point, and for the most part, doesn't matter all that much). Lastly, the caller already knows about the actual `k` through `num_documents`. So with all things considered, a single `float` is the the appropriate value.